### PR TITLE
BUG: ensure coordinate decs are always written out to tables.

### LINKED ISSUE #16998: Declination column missing when writing tables with SkyCoord columns to ECSV
### Description

If you create a new SkyCoord instance from an existing one, passing the `representation_type=` kwarg to the constructor, add it to a table, and write the table to ECSV, columns are missing.

For example, this code:

```python
coords = SkyCoord(np.zeros(5), np.ones(5), np.zeros(5), representation_type=CartesianRepresentation)
coords = SkyCoord(coords, representation_type=UnitSphericalRepresentation)
Table({"coords": coords}).write(sys.stdout, format="ascii.ecsv")
```

results in the following table output, missing the declination column:

```
# %ECSV 1.0
# ---
# datatype:
# - {name: coords.ra, unit: deg, datatype: float64}
# meta: !!omap
# - __serialized_columns__:
#     coords:
#       __class__: astropy.coordinates.sky_coordinate.SkyCoord
#       frame: icrs
#       ra: !astropy.table.SerializedColumn
#         __class__: astropy.coordinates.angles.core.Longitude
#         unit: &id001 !astropy.units.Unit {unit: deg}
#         value: !astropy.table.SerializedColumn {name: coords.ra}
#         wrap_angle: !astropy.coordinates.Angle
#           unit: *id001
#           value: 360.0
#       representation_type: unitspherical
# schema: astropy-2.0
coords.ra
90.0
90.0
90.0
90.0
90.0
```

### Expected behavior

Each of the three examples in the code below should print:

```
# %ECSV 1.0
# ---
# datatype:
# - {name: coords.ra, unit: deg, datatype: float64}
# - {name: coords.dec, unit: deg, datatype: float64}
# meta: !!omap
# - __serialized_columns__:
#     coords:
#       __class__: astropy.coordinates.sky_coordinate.SkyCoord
#       dec: !astropy.table.SerializedColumn
#         __class__: astropy.coordinates.angles.core.Latitude
#         unit: &id001 !astropy.units.Unit {unit: deg}
#         value: !astropy.table.SerializedColumn {name: coords.dec}
#       frame: icrs
#       ra: !astropy.table.SerializedColumn
#         __class__: astropy.coordinates.angles.core.Longitude
#         unit: *id001
#         value: !astropy.table.SerializedColumn {name: coords.ra}
#         wrap_angle: !astropy.coordinates.Angle
#           unit: *id001
#           value: 360.0
#       representation_type: spherical
# schema: astropy-2.0
coords.ra coords.dec
90.0 0.0
90.0 0.0
90.0 0.0
90.0 0.0
90.0 0.0
```

However, only the first two tables print that. The last table is missing the declination column.

### How to Reproduce

```python
from astropy import units as u
from astropy.coordinates import CartesianRepresentation, UnitSphericalRepresentation, SkyCoord
import numpy as np
from astropy.table import Table
import sys

coords = SkyCoord(np.full(5, 90), np.zeros(5), unit=u.deg)
Table({"coords": coords}).write(sys.stdout, format="ascii.ecsv")

coords = SkyCoord(np.zeros(5), np.ones(5), np.zeros(5), representation_type=CartesianRepresentation)
coords.representation_type = UnitSphericalRepresentation
Table({"coords": coords}).write(sys.stdout, format="ascii.ecsv")

coords = SkyCoord(np.zeros(5), np.ones(5), np.zeros(5), representation_type=CartesianRepresentation)
coords = SkyCoord(coords, representation_type=UnitSphericalRepresentation)
Table({"coords": coords}).write(sys.stdout, format="ascii.ecsv")
```

### Versions

```
Python 3.11.10 (main, Sep  7 2024, 05:47:53) [Clang 15.0.0 (clang-1500.1.0.2.5)]
astropy 6.1.3
Numpy 2.1.1
pyerfa 2.0.1.4
Scipy 1.14.1
Matplotlib 3.9.2
```

### PR DESCRIPTION
**Pull request by mhvk**
This pull request is to address a bug in how coordinates in tables are written out: if the data are unitspherical, the last representation name was removed, which only is OK for the (default) case in which the representation class is spherical. The corrects that, as well as similar problems for proper motion and radial-velocity.

In the process, also allows things like `ICRS(ra, dec, radial_velocity=..., differential_type="radial")` (which raised on `radial_velocity` before, since the default name was not set).

p.s. All this is a lot of hassle just because the frames insist on hiding the actual data inside different representations -- and then wish to write those for-show ones to files instead of using the true underlying data. But that's for another time... 

Fixes #16998

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
@mhvk , you milestoned to v7 but also added backport label. They are in conflict. Which one is correct? Hope you can clarify. Thanks!

**Comment by mhvk** [comment]
Oops, thanks for catching that. I thought it would be good to backport!

File: astropy/coordinates/baseframe.py
Comment (lpsinger) [inline astropy/coordinates/baseframe.py:203]: What does "they've been specific" mean?

**Review by lpsinger — CHANGES_REQUESTED** [review]
I suggest getting a second reviewer. I'm not an expert on the internals of astropy.coordinates.

File: astropy/coordinates/baseframe.py
Comment (mhvk) [reply astropy/coordinates/baseframe.py]: OK, thanks, corrected!

**Comment by mhvk** [comment]
For further review: the last commit only affected a comment, so I used `[skip ci]` - the previous commit passed all tests, so this should be fine. Do squash the two on merge!

File: astropy/coordinates/baseframe.py
Comment (eerovaher) [inline astropy/coordinates/baseframe.py:213]: Slicing a `list` forces making a copy:
```suggestion
                        diff_attrs.pop()
```

Comment (mhvk) [reply astropy/coordinates/baseframe.py:213]: I can do that, but prefer not to, as I think the code is clearer if it uses the slices consistently (and this is not performance critical).

**Review by eerovaher — COMMENTED** [review]
I don't like the regression tests because they invoke too much of the I/O machinery that is not really relevant to the problem and because they also explicitly use the private API. For `SkyCoord` we could create a table with the public `to_table()` method and check what that looks like, but the problem is that `BaseCoordinateFrame` does not have that method. When `to_table()` was implemented it was implemented only on `SkyCoord` because it was not even possible to use a `BaseCoordinateFrame` as a mixin column in a `Table` (that has been possible only since #16831). But I can't think of a good reason why there shouldn't be a `BaseCoordinateFrame.to_table()` method anymore. However, it is not obvious to me that `BaseCoordinateFrame.to_table()` should be a blocker for this bugfix, even if it would improve the regression tests a lot.

It would be best to submit bugfixes in two commits. The first commit should add the regression test for the bug, the second should fix the bug and add the change log entry. Checking out the first commit and running the tests would then very simply verify both that the bug exists and that the new test catches it, and it would then be clear that the second commit must fix the bug because it can't just hide the bug by editing the tests if it doesn't touch the tests at all. Git makes this workflow very simple because Git does not force you to make the commit with the regression test before you start working on the bugfix.

File: astropy/coordinates/tests/test_regression.py
Comment (mhvk) [reply astropy/coordinates/tests/test_regression.py:775]: I included the test of writing out the table since that was the one in the original issue... And this is fine in the sense that really `_represent_as_dict()` exists only for serializing a mixin column to a set of regular columns. But obviously, it is a bit silly to keep on creating tables when we know what the problem is, so I settled for making an exception to the testing of private methods, putting it in `test_regression` on purpose since I couldn't think of a nice, proper way to test...

Maybe two options: add a comment here along the lines of what I wrote above, or add a test column to `astropy/io/tests/mixin_columns.py`. Happy to go either way. 

Anyway, I basically agree with you that one should not test private

File: astropy/coordinates/baseframe.py
Comment (mhvk) [reply astropy/coordinates/baseframe.py]: OK, done

File: astropy/coordinates/baseframe.py
Comment (mhvk) [reply astropy/coordinates/baseframe.py]: Done

**Comment by mhvk** [comment]
I thought it really made most sense to test the table writing by adding a column to `mixin_columns`. Beautifully, that also brought out an extra bug, which is now fixed. I also used that opportunity to follow your suggestion of having two separate commits, one adding the tests, and the other adding the fixes. Much better now, thanks for the review!

File: astropy/coordinates/tests/test_regression.py
Comment (eerovaher) [inline astropy/coordinates/tests/test_regression.py:776]: The `get_kwargs()` helper method is too much complexity for a unit test. I'd prefer to store similar parameters together in dictionaries, which can be conveniently combined using the `|` operator. For example, see the definitions in https://github.com/astropy/astropy/blob/b266a4041ad4ebecbe1e07b329f70e03b71759e1/astropy/coordinates/tests/test_frames_with_velocity.py#L14-L17
and the usage throughout the file.

If these dictionaries are defined in the module scope then they can be reused in all the tests, but refactoring the already existing tests is beyond the scope of this pull request.

File: astropy/coordinates/tests/test_regression.py
Comment (eerovaher) [inline astropy/coordinates/tests/test_regression.py:784]: If you are already willing to let the tests generate `expected_entries` dynamically then why don't you replace the three test functions here with a single parametrized test? The big difference between this test and the other two is that in this case `expected_entries` does not contain `"differential_type"`, so it might be more convenient not to include this one in the parametrization, but combining the other two tests should be straightforward.

File: docs/changes/coordinates/16999.bugfix.rst
Comment (eerovaher) [inline docs/changes/coordinates/16999.bugfix.rst:2]:
```suggestion
the representation type was ``"unitspherical"``.
```

File: astropy/coordinates/tests/test_regression.py
Comment (mhvk) [reply astropy/coordinates/tests/test_regression.py:776]: I ended up doing it via parametrization, as you suggest below.

File: astropy/coordinates/tests/test_regression.py
Comment (mhvk) [reply astropy/coordinates/tests/test_regression.py:784]: You're right. I've now done it and it is a function again. Overall, perhaps a little less readable, but a lot shorter, so I think it is a win.

File: docs/changes/coordinates/16999.bugfix.rst
Comment (mhvk) [reply docs/changes/coordinates/16999.bugfix.rst:2]: Done

**Review by mhvk — COMMENTED** [review]
OK, that took me longer than perhaps was worth it, but it is better. Thanks!

**Comment by eerovaher** [comment]
Did you forget to push?

**Comment by mhvk** [comment]
I think this is all set - with `_represent_as_dict()` now hoped to become public (see #17015), I think it is fine to rely on it in the tests.

**Review by eerovaher — APPROVED** [review]
I still don't like that the regression tests are explicitly relying on private API and it would be better to rewrite them using the public  `BaseCoordinateFrame` and `SkyCoord` `to_table()` methods. There's just one problem - the pull request that implements `BaseCoordinateFrame.to_table()` has not been merged yet, and because it was opened after this pull request it wouldn't be fair to have it block this one.

It seems to me like the quickest way forward is to merge this pull request as it is and then rewrite the new regression tests in the other pull request. Currently the `SkyCoord.to_table()` only has doctests, so unit tests for `to_table()` should be written anyways and rewriting the tests from here isn't any more work than writing brand new tests from scratch.

**Comment by mhvk** [comment]
Thanks, @eerovaher! As I wrote above, I think it is fine to leave the tests as they are given #17015 - though I guess there's also nothing wrong with tests for your `.to_table()` to include a frame with a unitspherical representation!

**Comment by eerovaher** [comment]
I'm not sure whose job it is to do the manual backport if the pull request was opened by an `astropy` maintainer.

**Pull request merged**
